src/python/pants/engine/README.md

```diff
@@ -105,29 +111,32 @@ class TypedDatatype(datatype([('field_name', Exactly(str, int))])):
 ```
 
 Assigning a specific type to a field can be somewhat unidiomatic in Python, and may be unexpected or
-unnatural to use. Additionally, the engine already applies a form of implicit type checking by
-ensuring there is a unique path from subject to product when a product request is made. However,
-regardless of whether the object is created directly with type-checked fields or whether it's
-produced from a set of rules by the engine's dependency injection, it is extremely useful to
-formalize the assumptions made about the value of an object into a specific type, even if the type
-just wraps a single field. The `datatype()` function makes it simple and efficient to apply that
-strategy.
+unnatural to use. However, regardless of whether the object is created directly with type-checked
+fields or whether it's produced from a set of rules by the engine's dependency injection, it is
+extremely useful to formalize the assumptions made about the value of an object into a specific type,
+even if the type just wraps a single field. The `datatype()` function makes it simple and efficient
+to apply that strategy.
+
+### Gets and RootRules
 
-### Parameter selectors and Gets
+As demonstrated above, parameter selectors select `@rule` arguments in the context of a set of `Params`.
+But where do `Params` come from?
 
-As demonstrated above, parameter selectors select `@rule` inputs in the context of a particular
-`Subject` (and its `Variants`: discussed below). But it is frequently necessary to "change" the
-subject and request products for subjects other than the one that the `@rule` is running for.
+One source of `Params` is the root of a request, where a `Param` type that may be provided by a caller
+of the engine can be declared using a `RootRule`. Installing a `RootRule` is sometimes necessary to
+seal the rule graph in cases where a `Param` could only possibly be computed outside of the rule graph
+and then passed in.
```

Comment:
I don't understand this. I think it would help to have an example specific to the idea of a `RootRule`. My main confusion is what does "root" mean in "root of a request"?